common/include/pcl/common/fps_counter.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2014-, Open Perception Inc.
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/common/time.h>
+
+#include <functional>
+
+namespace pcl {
+
+struct FpsCounterDetails {
+
+  unsigned counter_type_;
+  unsigned reset_iter_;
+  double fps_;
+  unsigned count_;
+  double start_time_;
+  double duration_;
+  double last_reset_;
+  FpsCounterDetails(unsigned counter_type = 0, unsigned reset_iter = 10)
+  : counter_type_(counter_type)
+  , reset_iter_(reset_iter)
+  , start_time_(pcl::getTime())
+  , last_reset_(pcl::getTime())
+  {}
+};
+
+class FpsCounter {
+public:
+  FpsCounter() {}
+  FpsCounter(FpsCounterDetails& details) : details_(&details) {}
+
+  FpsCounterDetails* details_;
```

Comment:
What about having the start time and tick count as class members as well?
```cpp
struct ret { double fps = 0, duration = 0; };

void reset() {
  tick = start_time = 0;
}

auto tick() {
  if (tick == 0) { start_time = now(); }
  tick += 1;
  return tick;
}

auto getCount() const { return tick; }

ret getAvgFps() const {
  const auto delta_t = now() - start_time;
  const auto fps = tick/delta_t;
  return {fps, delta_t};
}
```

Benefits:
* since start_time is stored, we can still perform the rest every second feature
* we could skip storing the fps quite easily since 99% of the time, fps will change every call
* it might be able to simplify the logic in lines 82-93 (I think it does, but not very sure)
* I feel this has all the information required for our "customFpsCounters"

Reply:
This is indeed doable and elegant. However, I think the whole calling structure should be changed in this case. By making start_time and tick as class members, I believe we have to create an FpsCounter object and pass it to all the callbacks on which we want to count the fps.  From the previous

```cpp
 //fps_calc
void callback(){
     static pcl::FpsCounterDetails counter; 
      //sleep_for(1sec)
     pcl::simpleFpsCounter(counter,"callback");
  }
```
to something like
```cpp
void callback(/* ... */, pcl::simpleFpsCounter & counter){  
   ...
  counter.tick();
  }

...
pcl::simpleFpsCounter counter(10,"callback");
callback(/* ... */, counter);
...

```
Under the hood details have to be addressed but I believe they will follow directly from this structure.
Probably this is a better way to tackle the bad structure present at the moment. I don't know if changing the callbacks' call can be a bad idea or not, though. What do you think @kunaltyagi ?

Reply:
If we do this, then we'd have to make tick thread_safe. This does look nicer though. A simple addition we can make is:
```cpp
struct TimerTicker {
    /* implicit */ TimerTicker(Timer& timer): _timer(timer) {}
    ~TimerTicker() { _timer.tick(); }
};
```
This will allow the callback to tick easily even if there are multiple return points.

Reply:
Thanks again for the useful comments @kunaltyagi 
I've already started to re-write the class with the new structure. The design is to have the base class as the "type 0" counter, or the one most used, while the conditional (derived class) will enclose the other 3 types encountered by overriding the ```resetCondition()``` and ```printDetails()``` functions.

I understand the thread safety requirement but I don't get what the ```TimerTicker```struct you suggested implies. Could you please give me more details on it and how eventually would change the calling site? 

I will follow shortly with a commit hopefully (work permitting 😄)

Reply:
Basically, if we can ensure the lifetime of the timer (via shared_ptr or some other means), we can create a class which:
* has a reference to the original timer
* created by copy, and on destruction performs tick

<table>
<tr>
<th>
Before
</th>
<th>
After
</th>
</tr>
<tr>
<td  valign="top">

<pre lang="cpp">
void callback(Timer& timer) {
  if (!this->advanced) {
    // simple code
    timer.tick();  // manual
    return
  }
  // complicated logic
  timer.tick();  // don't forget here as well
  return;
}
</pre>
</td>
<td  valign="top">

<pre lang="cpp">
void callback(Timer& timer) {
  TimerHelper h(timer); 
  // will auto-call timer.tick in dtor
  if (!this->advanced) {
    // simple code
    return
  }
  // complicated logic
  return;
}
</pre>
</td>
</tr>
</table>

Reply:
I have rewritten the class as mentioned in my last message. I split the cases considered so far into a derived class to keep the most common one as the base class (in case I can only do the base one). If this structure is ok, I will proceed to do the tick thread safe, which for now remains "manual", by means of the struct you suggested if that was what you had in mind.